cloudconfig/podcfg/podcfg.go

```diff
@@ -0,0 +1,464 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package podcfg
+
+import (
+	"fmt"
+	"net"
+	"path"
+	"reflect"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/version"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/agent"
+	agenttools "github.com/juju/juju/agent/tools"
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/cloudconfig/instancecfg"
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/tags"
+	"github.com/juju/juju/juju/paths"
+	"github.com/juju/juju/state/multiwatcher"
+	coretools "github.com/juju/juju/tools"
+)
+
+var logger = loggo.GetLogger("juju.cloudconfig.podcfg")
+
+// ControllerPodConfig represents initialization information for a new juju caas controller pod.
+type ControllerPodConfig struct {
+	// Tags is a set of tags/labels to set on the Pod, if supported. This
+	// should be populated using the PodLabels method in this package.
+	Tags map[string]string
+
+	// Bootstrap contains bootstrap-specific configuration. If this is set,
+	// Controller must also be set.
+	Bootstrap *BootstrapConfig
+
+	// Controller contains controller-specific configuration. If this is
+	// set, then the instance will be configured as a controller pod.
+	Controller *ControllerConfig
+
+	// APIInfo holds the means for the new pod to communicate with the
+	// juju state API. Unless the new pod is running a controller (Controller is
+	// set), there must be at least one controller address supplied.
+	// The entity name must match that of the pod being started,
+	// or be empty when starting a controller.
+	APIInfo *api.Info
+
+	// ControllerTag identifies the controller.
+	ControllerTag names.ControllerTag
+
+	// MachineNonce is set at provisioning/bootstrap time and used to
+	// ensure the agent is running on the correct pod.
+	MachineNonce string
+
+	// tools is the list of juju tools used to install the Juju agent
+	// on the new pod. Each of the entries in the list must have
+	// identical versions and hashes, but may have different URLs.
+	tools coretools.List
+
+	// DataDir holds the directory that juju state will be put in the new
+	// instance.
+	DataDir string
+
+	// LogDir holds the directory that juju logs will be written to.
+	LogDir string
+
+	// MetricsSpoolDir represents the spool directory path, where all
+	// metrics are stored.
+	MetricsSpoolDir string
+
+	// Jobs holds what machine jobs to run.
+	Jobs []multiwatcher.MachineJob
+
+	// MachineId identifies the new machine.
+	MachineId string
+
+	// AgentEnvironment defines additional configuration variables to set in
+	// the pod agent config.
+	AgentEnvironment map[string]string
+
+	// Series represents the pod series.
+	Series string
```

Comment:
Do we need Series?

Reply:
we need Series to determine DataDir, LogDir etc.